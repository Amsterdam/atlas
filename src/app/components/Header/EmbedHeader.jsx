import React from 'react'
import './EmbedHeader.scss'

const EmbedHeader = ({ closeAction }) => {
  const [embedLink, setEmbedLink] = React.useState('')
  const [iframe, setIframe] = React.useState('')

  React.useEffect(() => {
    // Todo: figure out a better way to embed a map
    // (e.g. by only sharing locations / results instead of just any state of the map)
    // Now we need to check the URL of the iFrame, as this can change
    const iframeObject = window.document.getElementById('atlas-iframe-map')
    const setUrlFromIframe = () => {
      if (iframe) {
        const {
          location: { href },
        } = iframeObject.contentWindow
        setEmbedLink(href)
        setIframe(`<iframe width="500" height="400" src="${href}"></iframe>`)
      }
    }

    setUrlFromIframe()

    const iframeChecker = window.setInterval(setUrlFromIframe, 500)

    return function cleanup() {
      window.clearInterval(iframeChecker)
    }
  }, [])

  const selectField = event => event.target.select()

  return (
    <div className="u-grid">
      <div className="u-row">
        <div className="u-col-sm--3">
          <h1 className="c-embed-header__title">Embedversie</h1>
        </div>
        <div className="u-col-sm--9">
          <nav>
            <button type="button" onClick={closeAction} className="c-embed-header__close" />
          </nav>
        </div>
      </div>
      <form className="c-embed-header__form">
        <div className="u-row c-embed-header__form-row qa-embed-header-form-link">
          <label
            htmlFor="linkField"
            className="u-col-sm--2 c-embed-header__field-label qa-embed-header-form-label"
          >
            Link
          </label>
          <div className="u-col-sm--10 c-embed-header__field-value">
            <input
              id="linkField"
              className="c-embed-header-form__input qa-embed-header-form-input"
              type="text"
              onClick={selectField}
              onChange={e => {
                setEmbedLink(e.target.value)
              }}
              value={embedLink}
              autoCapitalize="off"
              autoCorrect="off"
              autoComplete="off"
              spellCheck="false"
              placeholder="Link"
            />
          </div>
        </div>
        <div className="u-row c-embed-header__form-row qa-embed-header-form-html">
          <label
            htmlFor="iframe-field"
            className="u-col-sm--2 c-embed-header__field-label qa-embed-header-form-label"
          >
            HTML-code
          </label>
          <div className="u-col-sm--10 c-embed-header__field-value">
            <input
              id="iframe-field"
              className="c-embed-header-form__input qa-embed-header-form-input"
              type="text"
              onClick={selectField}
              onChange={e => {
                setIframe(e.target.value)
              }}
              value={iframe}
              autoCapitalize="off"
              autoCorrect="off"
              autoComplete="off"
              spellCheck="false"
              placeholder="Link"
            />
          </div>
        </div>
      </form>
    </div>
  )
}

export default EmbedHeader
