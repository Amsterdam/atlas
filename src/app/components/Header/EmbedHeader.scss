@import "../../../shared/styles/config";

$embed-header-height: 211px;
$embed-header-padding: 46px;
$embed-header-padding-icon: 4px;

.c-embed-header {
  &__title {
    @include font-size($xl-header-font);
  }

  &__close {
    @include icon-button("icon-clear", $primary-dark, $primary-dark);

    background-color: $primary-light;
    position: absolute;
    right: -($embed-header-padding - $embed-header-padding-icon);
    top: -($embed-header-padding - $embed-header-padding-icon);

    &:hover,
    &:focus {
      background-color: $secondary-gray20;
    }
  }

  &-form__input {
    appearance: none;
    background-color: $primary-light;
    border: 1px solid $secondary-gray20;
    border-radius: 0;
    color: $primary-dark;
    outline: none;
    padding: ($base-whitespace / 2) $base-whitespace;
    width: 100%;

    &::placeholder {
      color: $secondary-gray60;
      font-weight: $normal-weight;
    }

    &:focus {
      border: 2px solid $primary-dark;
    }
  }

  &__form {
    margin-top: 24px;

    .c-embed-header__form-row {
      border-bottom: 1px solid $secondary-gray20;
      display: flex;
      margin: 0 -46px;
      padding: ($base-whitespace / 2) 0;
    }

    .c-embed-header__field-label {
      align-self: center;
      padding-left: $embed-header-padding;
    }

    .c-embed-header__field-value {
      align-self: center;
      padding-right: $embed-header-padding;
    }
  }
}
