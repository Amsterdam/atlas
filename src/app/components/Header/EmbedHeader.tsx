import { useState, useEffect } from 'react'
import { useDispatch } from 'react-redux'
import { Close } from '@amsterdam/asc-assets'
import styled from 'styled-components'
import { Button } from '@amsterdam/asc-ui'
import type { FunctionComponent, MouseEvent as ReactMouseEvent } from 'react'
import { hideEmbedPreview } from '../../../shared/ducks/ui/ui'

const CloseButton = styled(Button)`
  position: absolute;
  right: -42px;
  top: -42px;
`

const EmbedHeader: FunctionComponent = () => {
  const dispatch = useDispatch()
  const [embedLink, setEmbedLink] = useState('')
  const [iframe, setIframe] = useState('')

  useEffect(() => {
    // Todo: figure out a better way to embed a map
    // (e.g. by only sharing locations / results instead of just any state of the map)
    // Now we need to check the URL of the iFrame, as this can change
    const iframeObject = document.getElementById('atlas-iframe-map')
    const setUrlFromIframe = () => {
      if (iframeObject instanceof HTMLIFrameElement && iframeObject.contentWindow) {
        const {
          location: { href },
        } = iframeObject.contentWindow
        setEmbedLink(href)
        setIframe(`<iframe width="500" height="400" src="${href}"></iframe>`)
      }
    }

    setUrlFromIframe()

    const iframeChecker = window.setInterval(setUrlFromIframe, 500)

    return function cleanup() {
      window.clearInterval(iframeChecker)
    }
  }, [])

  const selectField = (event: ReactMouseEvent<HTMLInputElement>) => {
    if (event.target instanceof HTMLInputElement) {
      event.target.select()
    }
  }

  return (
    <div className="u-grid">
      <div className="u-row">
        <div className="u-col-sm--3">
          <h1 className="c-embed-header__title">Embedversie</h1>
        </div>
        <div className="u-col-sm--9">
          <nav>
            <CloseButton
              type="button"
              variant="blank"
              title="Sluiten"
              size={28}
              icon={<Close />}
              iconSize={15}
              onClick={() => dispatch(hideEmbedPreview())}
            />
          </nav>
        </div>
      </div>
      <form className="c-embed-header__form">
        <div className="u-row c-embed-header__form-row qa-embed-header-form-link">
          <label
            htmlFor="linkField"
            className="u-col-sm--2 c-embed-header__field-label qa-embed-header-form-label"
          >
            Link
          </label>
          <div className="u-col-sm--10 c-embed-header__field-value">
            <input
              id="linkField"
              className="c-embed-header-form__input qa-embed-header-form-input"
              type="text"
              onClick={selectField}
              onChange={(e) => {
                setEmbedLink(e.target.value)
              }}
              value={embedLink}
              autoCapitalize="off"
              autoCorrect="off"
              autoComplete="off"
              spellCheck="false"
              placeholder="Link"
            />
          </div>
        </div>
        <div className="u-row c-embed-header__form-row qa-embed-header-form-html">
          <label
            htmlFor="iframe-field"
            className="u-col-sm--2 c-embed-header__field-label qa-embed-header-form-label"
          >
            HTML-code
          </label>
          <div className="u-col-sm--10 c-embed-header__field-value">
            <input
              id="iframe-field"
              className="c-embed-header-form__input qa-embed-header-form-input"
              type="text"
              onClick={selectField}
              onChange={(e) => {
                setIframe(e.target.value)
              }}
              value={iframe}
              autoCapitalize="off"
              autoCorrect="off"
              autoComplete="off"
              spellCheck="false"
              placeholder="Link"
            />
          </div>
        </div>
      </form>
    </div>
  )
}

export default EmbedHeader
