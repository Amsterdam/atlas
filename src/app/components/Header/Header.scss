@import "../../../shared/styles/config";

.styled-header {
  & > div > header {
    padding: 0 10px;
  }

  & > div > header > h1 {
    font-weight: 600;
  }

  @media screen and (min-width: 1200.02px) {
    & > div > header {
      padding: 0;
    }
  }

  & > div > header > div {
    justify-content: space-between;
  }

  [aria-hidden="true"] {
    display: none;
  }

  .auto-suggest {
    width: 50%;
  }
}
