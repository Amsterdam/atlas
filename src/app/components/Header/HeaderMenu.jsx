import React from 'react'
import { MenuInline, MenuToggle, MenuFlyOut, MenuItem, MenuButton } from '@datapunt/asc-ui'
import { ChevronRight } from '@datapunt/asc-assets'
import PropTypes from 'prop-types'
import Link from 'redux-first-router-link'
import {
  toApisPage,
  toAvailabilityPage,
  toDatasets,
  toHelpPage,
  toMaintentancePage,
  toMap,
  toPanoramaAndPreserveQuery,
  toPrivacyPage,
} from '../../../store/redux-first-router/actions'
import truncateString from '../../../shared/services/truncateString/truncateString'

const toPanoramaAction = toPanoramaAndPreserveQuery(undefined, undefined, undefined, 'home')
const toMapAction = toMap(true)
const toDatasetsAction = toDatasets()
const toApisAction = toApisPage()
const toPrivacyAction = toPrivacyPage()
const toAvailabilityAction = toAvailabilityPage()
const toMaintentanceAction = toMaintentancePage()
const toHelpAction = toHelpPage()

const components = {
  default: MenuInline,
  mobile: MenuToggle,
}

const StyledLink = ({ children, ...otherProps }) => (
  <MenuButton $as={Link} {...otherProps}>
    {children}
  </MenuButton>
)

const HeaderMenu = ({ type, login, logout, user, showFeedbackForm, ...props }) => {
  const Menu = components[type]

  return (
    <Menu {...props}>
      <MenuFlyOut label="Onderdelen">
        <MenuItem>
          <StyledLink iconLeft={<ChevronRight />} to={toMapAction}>
            Kaart
          </StyledLink>
        </MenuItem>
        <MenuItem>
          <StyledLink iconLeft={<ChevronRight />} to={toPanoramaAction}>
            Panoramabeelden
          </StyledLink>
        </MenuItem>
        <MenuItem>
          <StyledLink iconLeft={<ChevronRight />} to={toDatasetsAction}>
            Datasets
          </StyledLink>
        </MenuItem>
        <MenuItem>
          <StyledLink iconLeft={<ChevronRight />} to={toApisAction}>
            Data services
          </StyledLink>
        </MenuItem>
      </MenuFlyOut>
      <MenuFlyOut label="Over">
        <MenuItem>
          <StyledLink iconLeft={<ChevronRight />} to={toPrivacyAction}>
            Privacy en informatiebeveiliging
          </StyledLink>
        </MenuItem>
        <MenuItem>
          <StyledLink iconLeft={<ChevronRight />} to={toAvailabilityAction}>
            Beschikbaarheid en kwaliteit data
          </StyledLink>
        </MenuItem>
        <MenuItem>
          <StyledLink iconLeft={<ChevronRight />} to={toMaintentanceAction}>
            Technisch beheer en werkwijze
          </StyledLink>
        </MenuItem>
        <MenuItem>
          <MenuButton iconLeft={<ChevronRight />} href="mailto:datapunt@amsterdam.nl">
            Contact
          </MenuButton>
        </MenuItem>
      </MenuFlyOut>
      <MenuItem>
        <MenuButton onClick={showFeedbackForm}>Feedback</MenuButton>
      </MenuItem>
      <MenuItem>
        <StyledLink to={toHelpAction}>Help</StyledLink>
      </MenuItem>

      {!user.authenticated ? (
        <MenuItem>
          <MenuButton onClick={login}>Inloggen</MenuButton>
        </MenuItem>
      ) : (
        <MenuFlyOut data-test="login" label={truncateString(user.name, 9)}>
          <MenuItem>
            <MenuButton onClick={logout} iconLeft={<ChevronRight />}>
              Uitloggen
            </MenuButton>
          </MenuItem>
        </MenuFlyOut>
      )}
    </Menu>
  )
}

HeaderMenu.propTypes = {
  login: PropTypes.func.isRequired,
  logout: PropTypes.func.isRequired,
  showFeedbackForm: PropTypes.func.isRequired,
  type: PropTypes.string.isRequired,
  user: PropTypes.shape({}).isRequired,
}

export default HeaderMenu
