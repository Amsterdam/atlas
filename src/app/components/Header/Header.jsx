import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import styled, { css } from 'styled-components'
import {
  AmsterdamLogo,
  Header as HeaderComponent,
  styles,
  breakpoint,
  showAboveBackDrop,
} from '@datapunt/asc-ui'
import HeaderSearchContainer from '../../../header/containers/header-search/HeaderSearchContainer'
import HeaderMenuContainer from './HeaderMenuContainer'

import EmbedHeader from './EmbedHeader'
import PrintHeader from './PrintHeader'

const stickyStyle = css`
  position: sticky;
  top: 0;
`

const HeaderWrapper = styled.section`
  width: 100%;

  // As position: sticky isn't supported on IE, this is needed to have position the header on top of the other content
  position: relative;

  ${showAboveBackDrop(true)}

  // Add position: sticky for supported browsers
  ${({ isHomePage }) =>
    isHomePage
      ? css`
          @media screen and ${breakpoint('max-width', 'laptopM')} {
            ${stickyStyle}
          }
        `
      : stickyStyle}
`

const StyledHeader = styled(HeaderComponent)`
  ${styles.HeaderNavigationStyle} {
    // This must be added to the @datapunt/asc-ui project https://github.com/Amsterdam/amsterdam-styled-components/issues/165
    @media screen and ${breakpoint('min-width', 'desktop')} {
      margin-left: 29px;
      margin-right: 29px;
    }

    @media screen and ${breakpoint('min-width', 'tabletM')} {
      justify-content: space-between;
    }
  }
`

const Header = ({
  homePage,
  printOrEmbedMode,
  printMode,
  embedPreviewMode,
  hasMaxWidth,
  hidePrintMode,
  hideEmbedMode,
}) => {
  if (!printOrEmbedMode) {
    return (
      <HeaderWrapper isHomePage={homePage} data-test="header">
        <StyledHeader
          tall={homePage}
          title="Data en informatie"
          homeLink="/"
          className="styled-header"
          fullWidth={!hasMaxWidth}
          logo={AmsterdamLogo}
          navigation={
            <>
              <HeaderSearchContainer />
              <HeaderMenuContainer
                type="default"
                data-test="header-menu-default"
                tall={homePage}
                showAt="laptopM"
              />
              <HeaderMenuContainer
                type="mobile"
                align="right"
                data-test="header-menu-mobile"
                hideAt="laptopM"
              />
            </>
          }
        />
      </HeaderWrapper>
    )
  }

  return (
    <div className={classNames({ 'u-fixed': !printMode && !embedPreviewMode })}>
      <div
        className={`c-dashboard__heading ${classNames({
          'o-max-width': hasMaxWidth,
        })}`}
      >
        <div className={classNames({ 'o-max-width__inner': hasMaxWidth })}>
          {printMode && <PrintHeader closeAction={hidePrintMode} />}
          {embedPreviewMode && <EmbedHeader closeAction={hideEmbedMode} />}
        </div>
      </div>
    </div>
  )
}

Header.propTypes = {
  homePage: PropTypes.bool.isRequired,
  hasMaxWidth: PropTypes.bool.isRequired,
  printMode: PropTypes.bool.isRequired,
  embedPreviewMode: PropTypes.bool.isRequired,
  printOrEmbedMode: PropTypes.bool.isRequired,
}

export default Header
