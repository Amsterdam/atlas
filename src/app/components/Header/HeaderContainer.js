import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { hideEmbedMode, hidePrintMode } from '../../../shared/ducks/ui/ui'
import Header from './Header'

/* istanbul ignore next */
const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      hidePrintMode,
      hideEmbedMode,
    },
    dispatch,
  )

export default connect(
  null,
  mapDispatchToProps,
)(Header)
