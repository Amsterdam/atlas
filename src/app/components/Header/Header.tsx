import {
  AmsterdamLogo,
  breakpoint,
  Header as HeaderComponent,
  showAboveBackDrop,
  styles,
} from '@amsterdam/asc-ui'
import classNames from 'classnames'
import { FunctionComponent } from 'react'
import styled, { css } from 'styled-components'
import HeaderSearch from '../../../header/components/HeaderSearch'
import EmbedHeader from './EmbedHeader'
import HeaderMenu from './HeaderMenu'
import PrintHeader from './PrintHeader'

const stickyStyle = css`
  position: sticky;
  top: 0;
`

interface HeaderWrapperProps {
  isHomePage: boolean
}

const HeaderWrapper = styled.section<HeaderWrapperProps>`
  width: 100%;
  // As position: sticky isn't supported on IE, this is needed to have position the header on top of the other content
  position: relative;

  ${({ isHomePage }) =>
    isHomePage
      ? css`
          @media screen and ${breakpoint('max-width', 'laptopM')} {
            ${stickyStyle}
          }
        `
      : stickyStyle}

  ${showAboveBackDrop(true) as any}
`

const StyledHeader = styled(HeaderComponent)`
  a {
    /* Making sure the anchors in the header have a decent clickable area size */
    display: flex;
    height: 100%;
    align-items: center;
  }

  ${styles.HeaderNavigationStyle} {
    // This must be added to the @amsterdam/asc-ui project https://github.com/Amsterdam/amsterdam-styled-components/issues/165
    @media screen and ${breakpoint('min-width', 'desktop')} {
      margin-left: 29px;
      margin-right: 29px;
    }

    @media screen and ${breakpoint('min-width', 'tabletM')} {
      justify-content: space-between;
    }
  }
`

export interface HeaderProps {
  homePage: boolean
  hasMaxWidth: boolean
  printMode: boolean
  embedPreviewMode: boolean
  printOrEmbedMode: boolean
}

const Header: FunctionComponent<HeaderProps> = ({
  homePage,
  printOrEmbedMode,
  printMode,
  embedPreviewMode,
  hasMaxWidth,
}) => {
  if (!printOrEmbedMode) {
    return (
      <HeaderWrapper isHomePage={homePage} data-test="header">
        <StyledHeader
          tall={homePage}
          title="Data en informatie"
          homeLink="/"
          className="styled-header"
          fullWidth={!hasMaxWidth}
          logo={AmsterdamLogo}
          navigation={
            <>
              <HeaderSearch />
              <HeaderMenu
                type="default"
                data-test="header-menu-default"
                tall={homePage}
                showAt="laptopM"
              />
              <HeaderMenu
                type="mobile"
                align="right"
                data-test="header-menu-mobile"
                hideAt="laptopM"
              />
            </>
          }
        />
      </HeaderWrapper>
    )
  }

  return (
    <div className={classNames({ 'u-fixed': !printMode && !embedPreviewMode })}>
      <div
        className={`c-dashboard__heading ${classNames({
          'o-max-width': hasMaxWidth,
        })}`}
      >
        <div className={classNames({ 'o-max-width__inner': hasMaxWidth })}>
          {printMode && <PrintHeader />}
          {embedPreviewMode && <EmbedHeader />}
        </div>
      </div>
    </div>
  )
}

export default Header
