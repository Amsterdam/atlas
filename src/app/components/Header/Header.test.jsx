import { shallow } from 'enzyme'
import React from 'react'
import Header from './Header'

// Mock the HeaderSearchContainer component as its not relevant for this test
jest.mock('../../../header/components/HeaderSearch', () => () => <div />)

describe('Header', () => {
  const props = {
    embedPreviewMode: false,
    hasEmbedButton: true,
    hasMaxWidth: true,
    hasPrintButton: true,
    homePage: false,
    printMode: false,
    printOrEmbedMode: false,
    user: {},
  }

  it('should render the main header', () => {
    const component = shallow(<Header {...props} />)
    expect(component).toMatchSnapshot()
  })

  it('should render the print header', () => {
    const component = shallow(
      <Header
        {...{
          ...props,
          printOrEmbedMode: true,
          printMode: true,
        }}
      />,
    )
    expect(component).toMatchSnapshot()
  })

  it('should render the embed header', () => {
    const component = shallow(
      <Header
        {...{
          ...props,
          printOrEmbedMode: true,
          embedPreviewMode: true,
        }}
      />,
    )
    expect(component).toMatchSnapshot()
  })
})
