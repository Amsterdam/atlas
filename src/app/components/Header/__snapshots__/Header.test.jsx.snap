// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render the embed header 1`] = `
<div
  className=""
>
  <div
    className="c-dashboard__heading o-max-width"
  >
    <div
      className="o-max-width__inner"
    >
      <EmbedHeader />
    </div>
  </div>
</div>
`;

exports[`Header should render the main header 1`] = `
<styled.section
  data-test="header"
  isHomePage={false}
>
  <Styled(Header)
    className="styled-header"
    fullWidth={false}
    homeLink="/"
    logo={[Function]}
    navigation={
      <React.Fragment>
        <Memo(Connect(HeaderSearch)) />
        <Memo(Connect(HeaderMenu))
          data-test="header-menu-default"
          showAt="laptopM"
          tall={false}
          type="default"
        />
        <Memo(Connect(HeaderMenu))
          align="right"
          data-test="header-menu-mobile"
          hideAt="laptopM"
          type="mobile"
        />
      </React.Fragment>
    }
    tall={false}
    title="Data en informatie"
  />
</styled.section>
`;

exports[`Header should render the print header 1`] = `
<div
  className=""
>
  <div
    className="c-dashboard__heading o-max-width"
  >
    <div
      className="o-max-width__inner"
    >
      <PrintHeader />
    </div>
  </div>
</div>
`;
